rasterio/_features.pyx

```diff
@@ -549,6 +549,10 @@ cdef class GeomBuilder:
         coordinates = [p['coordinates'] for p in self._buildParts(self.geom)]
         return {'type': 'MultiPolygon', 'coordinates': coordinates}
 
+    cpdef _buildGeometryCollection(self):
+        geometries = [geom for geom in self._buildParts(self.geom)]
+        return {'type': 'GeometryCollection', 'geometries': geometries}
```

Comment:
This is where support was added. OGRGeomBuilder supports reading geometry collections and this is essentially the inverse of that logic.